cvmfs/upload_gateway.cc

```diff
@@ -0,0 +1,215 @@
+    LogCvmfs(kLogUploadGateway, kLogStderr,
+             "Failed to configure HTTP uploader. "
+             "Missing session token file.\n");
+    return false;
+  }
+  config->session_token_file = spooler_definition.session_token_file;
+
+  // Repo address, e.g. http://my.repo.address
+  config->api_url = config_string;
+
+  return true;
+}
+
+GatewayUploader::GatewayUploader(const SpoolerDefinition& spooler_definition)
+    : AbstractUploader(spooler_definition),
+      config_(),
+      session_context_(new SessionContext()) {
+  assert(spooler_definition.IsValid() &&
+         spooler_definition.driver_type == SpoolerDefinition::Gateway);
+
+  if (!ParseSpoolerDefinition(spooler_definition, &config_)) {
+    abort();
+  }
+
+  atomic_init32(&num_errors_);
+
+  LogCvmfs(kLogUploadGateway, kLogStderr,
+           "HTTP uploader configuration:\n"
+           "  API URL: %s\n"
+           "  Session token file: %s\n",
+           config_.api_url.c_str(), config_.session_token_file.c_str());
+}
+
+GatewayUploader::~GatewayUploader() {
+  if (session_context_) {
+    delete session_context_;
+  }
+}
+
+bool GatewayUploader::Initialize() {
+  if (!AbstractUploader::Initialize()) {
+    return false;
+  }
+  std::string session_token;
+  if (!ReadSessionTokenFile(config_.session_token_file, &session_token)) {
+    return false;
+  }
+  return session_context_->Initialize(config_.api_url, session_token);
+}
+
+bool GatewayUploader::FinalizeSession() { return session_context_->Finalize(); }
+
+void GatewayUploader::WaitForUpload() const {
+  session_context_->WaitForUpload();
+}
+
+std::string GatewayUploader::name() const { return "HTTP"; }
+
+bool GatewayUploader::Remove(const std::string& /*file_to_delete*/) {
+  return false;
+}
+
+bool GatewayUploader::Peek(const std::string& /*path*/) const { return false; }
+
+bool GatewayUploader::PlaceBootstrappingShortcut(
+    const shash::Any& /*object*/) const {
+  return false;
+}
+
+unsigned int GatewayUploader::GetNumberOfErrors() const {
+  return atomic_read32(&num_errors_);
+}
+
+void GatewayUploader::FileUpload(const std::string& local_path,
+                                 const std::string& remote_path,
+                                 const CallbackTN* callback) {
+  UniquePtr<GatewayStreamHandle> handle(
+      new GatewayStreamHandle(callback, session_context_->NewBucket()));
+
+  FILE* local_file = fopen(local_path.c_str(), "rb");
+  if (!local_file) {
+    LogCvmfs(kLogUploadGateway, kLogStderr,
+             "File upload - could not open local file.");
+    BumpErrors();
+    Respond(callback, UploaderResults(1, local_path));
+    return;
+  }
+
+  std::vector<char> buf(1024);
+  size_t read_bytes = 0;
+  do {
+    read_bytes = fread(&buf[0], buf.size(), 1, local_file);
+    ObjectPack::AddToBucket(&buf[0], buf.size(), handle->bucket);
+  } while (read_bytes == buf.size());
+  fclose(local_file);
+
+  shash::Any content_hash(shash::kSha1);
```

Comment:
We might rather want to use `spooler_definition().hash_algorithm` from the parent `AbstractUploader` class.